packages/form-control/src/FormControlMixin.ts

```diff
@@ -0,0 +1,528 @@
+import { IElementInternals } from 'element-internals-polyfill';
+import { Constructor, FormControlInterface, IControlHost, Validator } from './types';
+
+export function FormControlMixin<T extends Constructor<HTMLElement & IControlHost>>(SuperClass: T) {
+  class FormControl extends SuperClass {
+    /** Wires up control instances to be form associated */
+    static get formAssociated() {
+      return true;
+    }
+
+    /**
+     * A list of Validator objects that will be evaluated when a control's form
+     * value is modified or optionally when a given attribute changes.
+     *
+     * When a Validator's callback returns false, the entire form control will
+     * be set to an invalid state.
+     */
+    static get formControlValidators(): Validator[] {
+      return [];
+    }
+
+    /**
+     * Allows the FormControl instance to respond to Validator attributes.
+     * For instance, if a given Validator has a `required` attribute, that
+     * validator will be evaluated whenever the host's required attribute
+     * is updated.
+     */
+    static get observedAttributes(): string[] {
+      const validatorAttributes = this.formControlValidators
+        .map(validator => validator.attribute);
+
+      /** @ts-ignore This exits */
+      const observedAttributes = super.observedAttributes || [];
+
+      /** Make sure there are no duplicates inside the attributes list */
+      const attributeSet = new Set([...observedAttributes, ...validatorAttributes])
+      return [...attributeSet];
+    }
+
+    /**
+     * Return the validator associated with a given attribute. If no
+     * Validator is associated with the attribute, it will return null.
+     */
+    static getValidator(attribute: string): Validator {
+      return this.formControlValidators.find(validator =>
+        validator.attribute === attribute
+      );
+    }
+
+    /** The ElementInternals instance for the control. */
+    internals = this.attachInternals() as unknown as IElementInternals;
+
+    /** Keep track of if the control has focus */
+    focused = false;
```

Comment:
Some state you track privately, but this you're tracking publicly, what's the benefit here? 

Unrelated to the technical implementation, I ask due to the fact that many elements I use already leverage this property to proxy `:focus-visible`.

Reply:
Yeah, this is mostly a convenience thing. It might be the case that a user wants their own show error algorithm. I suppose I should let them take care of that, however.